developing/languages/GO.md

```diff
@@ -4,9 +4,184 @@
+This guide provides some general resources for working in Go.
+[Web Application in Go](../../web/server/go.md)
+provides some specifics to web development.
+
+## Learning Resources
+
+### References
 
 * [The Go Programming Language Official Website](https://golang.org/)
+* [Effective Go](https://golang.org/doc/effective_go.html) (how to do things “the Go way”)
 * [Go Pointer Primer](https://github.com/trussworks/go-pointer-primer)
+* [GoDoc](https://godoc.org/) (where you can read the docs for nearly any Go package)
+* [Go wiki](https://github.com/golang/go/wiki/Learn)
+* _Book_: [The Go Programming Language](http://www.gopl.io/)
+* Advanced Testing with Go
+  [Video](https://www.youtube.com/watch?v=yszygk1cpEc)
+  and [Article](https://about.sourcegraph.com/go/advanced-testing-in-go) (great overview of useful techniques, useful for all Go programmers)
+
+### Tours/Lessons
+
+* [A Tour of Go](https://tour.golang.org) (in-browser interactive language tutorial)
+* [How to Write Go Code](https://golang.org/doc/code.html) (info about the Go environment, testing, etc.)
 * [Go by Example](https://gobyexample.com)
 * [Go Track on Exercism](https://exercism.io/tracks/go)
+
+## Testing
+
+### General
+
+* Use table-driven tests where appropriate.
+* Make judicious use of helper functions so that the intent of a test is not lost in a sea of error checking and boilerplate.
+  Use [`t.Helper()`](https://golang.org/pkg/testing/#T.Helper) in your test helper functions to keep stack traces clean.
+
+### Coverage
+
+* Always test exported functions.
+  Exported functions should be treated as an API layer for other packages.
+  Cover the expected behavior and error scenarios as a user of that API.
+* Try not to test unexported functions.
+  Unexported functions are implementation details of exported ones
+  and should not change the intended usage.
+  If you find that an unexported function is complex and needs testing,
+  it might mean it needs to be refactored as it's exported function elsewhere.
+
+## Packages
+
+### Prefer Standard Libraries
+
+In general,
+when selecting new packages,
+highly consider standard libraries over third party dependencies.
+One of the strengths of Go
+is its core packages,
+such as
+[http](https://golang.org/pkg/net/http/),
+[json](https://golang.org/pkg/encoding/json/),
+and [sql](https://golang.org/pkg/database/sql/).
+These libraries also use vocabulary and patterns
+easily accessible via popular public Go resources,
+which are often translatable to modern programming approaches
+in their respective areas.
+This creates an easier bridge
+for Engineers new to Go
+or domain areas (such as relational databases)
+to adjust and onboard.
+
+Since the third party ecosystem is still new,
+many packages tend to be immature with little community support,
+follow opinionated patterns inconsistent with Go idioms,
+or lack long term support.
+Be aware that choosing a new package
+is likely to involve contributing to its upkeep.
```

Comment:
While I understand the concern here, this feels a little too opinionated to me toward the side of "only use the standard library".  I'd hate for us to get into a "not invented here" syndrome with this.  I think the big takeaway is that one should exercise caution in choosing third-party dependencies (with any language, but perhaps more so in Go), and weigh the risks you mentioned against the benefits.  For what it's worth, there was some [discussion about this topic](https://changelog.com/gotime/80#transcript-112) in a recent Go Time podcast.

Reply:
I don't think that someone is likely to use a third-party JSON library at this point, but they might want to use a higher-level database abstraction than the built-in `database/sql`. How to we describe this recommendation in such a way that it can address that situation?

Reply:
Add language to the effect of: if you do use a third-party library, be prepared to explain why the standard library couldn't meet our needs, and also why you think the risks of a third-party library are justified or mitigated in this case.

Reply:
Going to change the language around here to soften it a bit.